### ISSUE TITLE
DNS rewriting issue

### ISSUE THREAD

**Issue by liuhome11** [AdguardTeam/AdGuardHome#6667]
### Prerequisites

- [X] I have checked the [Wiki](https://github.com/AdguardTeam/AdGuardHome/wiki) and [Discussions](https://github.com/AdguardTeam/AdGuardHome/discussions) and found no answer

- [X] I have searched other issues and found no duplicates

- [X] I want to request a feature or enhancement and not ask a question

### The problem

How can we configure multiple IP addresses to enhance reliability for DNS rewriting, as the current setup only supports a single IP address and raises concerns about a single point of failure?

### Proposed solution

steamcommunity.com
45.114.11.25,118.193.240.37,210.56.51.192

**Comment by liuhome11** [comment]
||aa.xxx.net^$dnsrewrite=NOERROR;A;123.123.123.123
||aa.xxx.net^$dnsrewrite=NOERROR;A;21.21.21.21                                                                                                                          
    After testing with the mentioned configuration, it was observed that only the initial IP is matched successfully. In the event of a failure in the first IP, it does not automatically switch to the subsequent parameters. Checking the resolution using 'nslookup -displaydns' reveals that the second IP, as configured, is present in the resolution.
     The ability to swiftly replace the IP address assigned by CDN in the event of a failure, or to implement DNS rewriting protection features such as polling, can be achieved through various means.

**Comment by liuhome11** [comment]
![image](https://github.com/AdguardTeam/AdGuardHome/assets/157213077/ac127cb5-af50-4b27-b176-734ee57f1272)
![image](https://github.com/AdguardTeam/AdGuardHome/assets/157213077/2f25fb15-1ba9-4075-8b3f-f011f716480b)
I would like to support single/multiple domain rewriting with single/multiple IP addresses. Many CDN websites have similar requirements, and I hope to have a better way to achieve this, such as using the 'host' method. Thank you

**Comment by ainar-g** [comment]
I'm still not exactly sure what you're asking.  Your example with two `dnsrewrite` rules works:

```sh
dig 'aa.xxx.net' IN A
```

```none
[…]
;; QUESTION SECTION:
;aa.xxx.net.                    IN      A

;; ANSWER SECTION:
aa.xxx.net.             10      IN      A       123.123.123.123
aa.xxx.net.             10      IN      A       21.21.21.21
[…]
```

**Comment by liuhome11** [comment]
@ainar-g I want to convey that a specific domain is bound to multiple IP addresses, and when a user makes a request, it randomly matches one IP from the list of bindings for resolution. For example: AA.XXX.COM is bound to 45.114.11.25, 118.193.240.37, 210.56.51.192. When a user requests, ADG randomly matches one IP, such as 210.56.51.192, for the end user.

**Comment by ainar-g** [comment]
If I understood correctly, there are basically two feature requests here.  One is for the legacy rewrite UI to support multiple targets and responses.  I don't think this is likely, since that makes the feature more complicated, but we might consider something like that in the future.

The other is randomization of the response records from `$dnsrewrite` rules (and other types of rewrites?).  I'm not sure if we need that, since it is a client's choice which of the records to choose, most of the time.

**Comment by liuhome11** [comment]
> 如果我理解正确的话，这里基本上有两个功能请求。一种是让旧版重写 UI 支持多个目标和响应。我认为这不太可能，因为这会使功能更加复杂，但我们可能会在未来考虑类似的事情。
> 
> 另一种是来自规则的响应记录的随机化（以及其他类型的重写？我不确定我们是否需要它，因为大多数时候选择哪条记录是客户的选择。`$dnsrewrite`

This requirement is for certain software or programs, such as Steam, which, in the Chinese environment, relies on CDN services. It needs to be provided with correct and uncontaminated IP addresses. Simultaneously, to prevent issues in case a single IP fails, measures should be taken to ensure continuous accessibility.

**Comment by ainar-g** [comment]
That could be achieved by making the *Fastest IP Address* upstream mode also ping addresses from rewrites.  Which is an interesting but probably separate feature.

**Comment by liuhome11** [comment]
> 这可以通过使_最快的 IP_ 地址上行模式也对重写的地址执行 ping 来实现。这是一个有趣但可能是独立的功能。

In my practical application, I only permit the resolution of the website to specific IP addresses, binding these fixed IPs, and utilizing VPN for traffic diversion to prevent a large volume of unrelated applications from occupying the VPN channel bandwidth.

**Comment by narqo** [comment]
> One is for the legacy rewrite UI to support multiple targets and responses. I don't think this is likely, since that makes the feature more complicated, but we might consider something like that in the future.

I have a relatively similar request. That is, I use adguard as a local DNS server in my home network. I added a custom A record in the DNS rewrites to resolve `*.pi.home` into an IP of the home lab k8s cluster.

```
% dig test.pi.home

;; QUESTION SECTION:
;test.pi.home.			IN	A

;; ANSWER SECTION:
test.pi.home.		10	IN	A	192.168.20.43
```

Because the cluster runs on multiple nodes, each node could serve the ingress requests. Thus, the DNS rewrite could technically return multiple IPs. Unfortunately, this isn't seem possible.

**Comment by mflagg2814** [comment]
I'd also like to have a DNS rewrite for an A record return multiple IPs.

**Comment by tommyalatalo** [comment]
Same here, it would be very useful to set multiple IP numbers to the record when managing HA setups for instance.

**Comment by NikMAX2077** [comment]
Same here. 
I'm using ADH as local DNS for my home network. When I make a DNS query with one of the `*.home` domains (e.g `container.home`), it's redirected to the Nginx Proxy Manager (e.g `192.168.1.100`) which in turn redirects it to the target docker container (e.g `container-name:port`). It works great, but only inside my local network.
I want to use this setup both with my home local network (e.g `192.168.x.x`) and tailscale (e.g. `100.x.x.x`), but now I'm forced to choose only one thing.
If I creates two rewrites with the same domain and different ips only the first one will works. For example:
```
*.home -> 192.168.1.100
*.home -> 100.1.2.3
```
This configuration returns the right route when I'm perform `http://test.home` from my home local network. But when I'm trying to perform `http://test.home` from my friend's house using tailscale ADH still return the same `192.168.1.100` that points to friend's local network, not mine.
Swap records and place tailscale first It's not helpful too:
```
*.home -> 100.1.2.3
*.home -> 192.168.1.100
```
In that case ADH always returns 100.1.2.3 that allow me to connect to my containers by domain names only when Tailscale enabled on my phone, even when I'm home.


To summarize: in cases where multiple rewrites are created for the same domain name, it would be very useful for ADH to sequentially check them all and return only the one that works. Or we need the ability to return different ip depending on the client's ip or client's network type.
